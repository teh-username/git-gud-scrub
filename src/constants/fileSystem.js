export default {
  MAX_NUMBER_OF_FILES: 5,
};
