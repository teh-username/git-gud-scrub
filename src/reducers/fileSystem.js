import { combineReducers } from 'redux';

export const initialState = {
  files: ['woof.py', 'ruff.py'],
  fileStatus: {
    'woof.py': {
      tracked: true,
      modified: true,
      staged: true,
    },
    'ruff.py': {
      tracked: true,
      modified: true,
      staged: true,
    },
  },
};

const files = (state = initialState.files, action) => {
  return state;
};
const fileStatus = (state = initialState.fileStatus, action) => {
  return state;
};

export default combineReducers({
  files,
  fileStatus,
});

export const getFiles = ({ fileSystem }) => fileSystem.files;
export const getFileStatus = ({ fileSystem }) => fileSystem.fileStatus;
