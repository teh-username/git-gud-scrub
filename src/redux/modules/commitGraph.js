import { combineReducers } from 'redux';
import {
  GIT_BRANCH_ADD,
  GIT_BRANCH_DELETE,
  GIT_BRANCH_CHECKOUT,
  GIT_COMMIT,
} from './gitEmulator';
import grapher from '../../utils/grapher';

const initialState = {
  head: 'qa9j',
  branches: {
    master: 'qa9j',
    derp: 'kgxu',
    nope: 'wnrw',
  },
  commits: ['ko9z', 'plve', 'cb54', 'wnrw', 'kgxu', 'qa9j'],
  lookup: {
    ko9z: {
      message: 'Lorem ipsum dolor sit amet',
      children: ['plve', 'cb54'],
    },
    plve: {
      message: 'consectetur adipiscing elit',
      children: ['wnrw'],
    },
    cb54: {
      message: 'Donec molestie dictum dolor',
      children: ['kgxu', 'qa9j'],
    },
    wnrw: {
      message: 'Curabitur maximus ante velit',
      children: [],
    },
    kgxu: {
      message: ' Aenean sit amet enim at tellus vulputate consectetur',
      children: [],
    },
    qa9j: {
      message: 'Nullam auctor sapien a laoreet luctus',
      children: [],
    },
  },
};

const lookup = (state = initialState.lookup, action) => {
  switch (action.type) {
    case GIT_COMMIT:
      return {
        ...state,
        [action.parentRef]: {
          ...state[action.parentRef],
          children: [...state[action.parentRef].children, action.ref],
        },
        [action.ref]: {
          message: action.message,
          children: [],
        },
      };
    default:
      return state;
  }
};
const commits = (state = initialState.commits, action) => {
  switch (action.type) {
    case GIT_COMMIT:
      return [...state, action.ref];
    default:
      return state;
  }
};
const branches = (state = initialState.branches, action) => {
  switch (action.type) {
    case GIT_BRANCH_ADD:
      return {
        ...state,
        [action.name]: action.ref,
      };
    case GIT_BRANCH_DELETE:
      return Object.entries(state)
        .filter(([branch, ref]) => branch !== action.name)
        .reduce(
          (acc, [branch, ref]) => ({
            ...acc,
            [branch]: ref,
          }),
          {}
        );
    case GIT_COMMIT:
      return Object.entries(state)
        .filter(([branch, ref]) => ref === action.parentRef)
        .reduce(
          (acc, [branch, _]) => ({
            ...acc,
            [branch]: action.ref,
          }),
          { ...state }
        );
    default:
      return state;
  }
};
const head = (state = initialState.head, action) => {
  switch (action.type) {
    case GIT_BRANCH_CHECKOUT:
    case GIT_COMMIT:
      return action.ref;
    default:
      return state;
  }
};

export default combineReducers({
  lookup,
  commits,
  branches,
  head,
});

export const getGraphFromState = ({
  commitGraph: { lookup, commits, branches, head },
}) =>
  grapher({
    lookup,
    commits,
    branches,
    head,
  });

export const getHeadReference = ({ commitGraph: { head } }) => head;
export const getBranchReference = ({ commitGraph: { branches } }, branch) =>
  branches[branch];
export const getBranches = ({ commitGraph: { branches } }) =>
  Object.keys(branches);
