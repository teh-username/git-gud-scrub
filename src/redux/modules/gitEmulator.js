export const GIT_ADD = 'modules/gitEmulator/GIT_ADD';

export const gitAdd = fileName => ({
  type: GIT_ADD,
  fileName
});

const gitCommandActionLookup = {
  add: gitAdd
};

export const executeCommand = input => dispatch => {
  // TODO: add check if command does not exist
  const { command, argument, flags } = input;
  const gitCommand = gitCommandActionLookup[command];
  dispatch(gitCommand(argument, flags));
};
