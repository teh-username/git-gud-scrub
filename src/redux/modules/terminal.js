import { combineReducers } from 'redux';

const ADD_LOG_INFO = 'modules/terminal/ADD_LOG_INFO';
export const ADD_LOG_ERROR = 'modules/terminal/ADD_LOG_ERROR';

/*
  Sample state:
  {
    entries: [
      {
        text: 'git status',
        logLevel: 'info',
      },
      {
        text: 'wat',
        logLevel: 'error',
      },
    ]
  }
*/
export const initialState = {
  entries: [],
};

export const entries = (state = initialState.entries, action) => {
  switch (action.type) {
    case ADD_LOG_INFO:
      return [...state, { text: action.text, logLevel: 'info' }];
    case ADD_LOG_ERROR:
      return [...state, { text: action.text, logLevel: 'error' }];
    default:
      return state;
  }
};

export default combineReducers({
  entries,
});

export const addLogInfo = text => ({
  type: ADD_LOG_INFO,
  text,
});

export const addLogError = text => ({
  type: ADD_LOG_ERROR,
  text,
});

export const getEntries = ({ terminal }) => terminal.entries;
