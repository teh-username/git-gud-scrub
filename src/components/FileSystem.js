import React from 'react';
import constants from '../constants/fileSystem';

const File = ({ name, statuses }) => (
  <li>
    <span>{name}</span>
    {!statuses.modified && (
      <button className="btn btn-link btn-sm">Modify</button>
    )}
  </li>
);

const FileList = ({ files, fileStatus }) => (
  <ul>
    {files.map(name => (
      <File key={name} name={name} statuses={fileStatus[name]} />
    ))}
  </ul>
);

class NewFileForm extends React.Component {
  render() {
    return (
      <div>
        <div className="input-group">
          <input type="text" className="form-control" placeholder="Filename" />
          <span className="input-group-btn">
            <button className="btn btn-secondary" type="button">
              Add File
            </button>
          </span>
        </div>
        <DuplicateFileNameWarning />
      </div>
    );
  }
}

const DuplicateFileNameWarning = () => (
  <div className="alert alert-warning" role="alert">
    <strong>Warning!</strong> That filename is already taken bud.
  </div>
);

const FileSystem = ({ files, fileStatus }) => (
  <div className="col-lg-3 col-sm-12">
    <h4>Files (Max of {constants.MAX_NUMBER_OF_FILES})</h4>
    <FileList files={files} fileStatus={fileStatus} />
    <NewFileForm />
  </div>
);

export default FileSystem;
