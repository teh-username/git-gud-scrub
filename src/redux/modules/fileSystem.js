import { combineReducers } from 'redux';
import { GIT_ADD, GIT_COMMIT } from './gitEmulator';

export const MAX_FILE_COUNT = 5;
const ADD_FILE = 'modules/fileSystem/ADD_FILE';
const MODIFY_FILE = 'modules/fileSystem/MODIFY_FILE';

/*
  Sample state:
  {
    files: ['woof.py', 'ruff.py'],
    fileStatus: {
      'woof.py': {
        tracked: true,
        modified: true,
        staged: true,
      },
      'ruff.py': {
        tracked: true,
        modified: true,
        staged: true,
      },
    },
  };
*/

export const initialState = {
  files: [],
  fileStatus: {},
};

export const files = (state = initialState.files, action) => {
  switch (action.type) {
    case ADD_FILE:
      return [...state, action.fileName];
    default:
      return state;
  }
};

export const fileStatus = (state = initialState.fileStatus, action) => {
  switch (action.type) {
    case ADD_FILE:
      return {
        ...state,
        [action.fileName]: {
          tracked: false,
          modified: undefined,
          staged: undefined,
        },
      };
    case MODIFY_FILE:
      return {
        ...state,
        [action.fileName]: {
          ...state[action.fileName],
          modified: true,
        },
      };
    case GIT_ADD:
      return {
        ...state,
        [action.fileName]: {
          ...state[action.fileName],
          tracked: true,
          staged: true,
          modified: false,
        },
      };
    case GIT_COMMIT:
      return Object.entries(state)
        .filter(([fileName, statuses]) => statuses.staged)
        .reduce((previousState, [fileName, statuses]) => {
          return {
            ...previousState,
            [fileName]: {
              ...previousState[fileName],
              staged: false,
            },
          };
        }, state);
    default:
      return state;
  }
};

export default combineReducers({
  files,
  fileStatus,
});

export const addFile = fileName => ({
  type: ADD_FILE,
  fileName,
});

export const modifyFile = fileName => ({
  type: MODIFY_FILE,
  fileName,
});

export const getFiles = ({ fileSystem }) => fileSystem.files;
export const getFileStatus = ({ fileSystem }) => fileSystem.fileStatus;
