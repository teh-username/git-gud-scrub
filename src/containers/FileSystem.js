import { connect } from 'react-redux';

import FileSystem from '../components/FileSystem';
import { getFiles, getFileStatus, addFile, modifyFile } from '../redux/modules/fileSystem';

const mapStateToProps = state => ({
  files: getFiles(state),
  fileStatus: getFileStatus(state),
});

export default connect(mapStateToProps, { addFile, modifyFile })(FileSystem);
