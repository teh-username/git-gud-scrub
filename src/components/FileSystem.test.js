import React from 'react';
import ReactDOM from 'react-dom';
import FileSystem, { File, FileList } from './FileSystem';
import { initialState } from '../redux/modules/fileSystem';

describe('FileSystem Component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<FileSystem {...initialState} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('File Component', () => {
    it('renders a unmodified file entry correctly', () => {
      const wrapper = shallow(
        <File name="heyo.js" statuses={{ modified: false }} />
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('renders a modified file entry correctly', () => {
      const wrapper = shallow(
        <File name="heyo.js" statuses={{ modified: true }} />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('FileList Component', () => {
    it('renders the correct number of File entries', () => {
      const wrapper = shallow(
        <FileList
          files={['heyo.js', 'heya.js']}
          fileStatus={{
            'heyo.js': { modified: true },
            'heya.js': { modified: false },
          }}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
