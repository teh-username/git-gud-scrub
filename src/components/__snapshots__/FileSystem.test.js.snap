// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileSystem Component DuplicateFileNameWarning should render correctly 1`] = `
<div
  className="alert alert-warning"
  role="alert"
>
  <strong>
    Warning!
  </strong>
   That filename is already taken bud.
</div>
`;

exports[`FileSystem Component File Component renders a modified file entry correctly 1`] = `
<li>
  <span>
    heyo.js
  </span>
</li>
`;

exports[`FileSystem Component File Component renders a unmodified file entry correctly 1`] = `
<li>
  <span>
    heyo.js
  </span>
  <button
    className="btn btn-link btn-sm"
  >
    Modify
  </button>
</li>
`;

exports[`FileSystem Component FileList Component renders the correct number of File entries 1`] = `
<ul>
  <Component
    key="heyo.js"
    name="heyo.js"
    statuses={
      Object {
        "modified": true,
      }
    }
  />
  <Component
    key="heya.js"
    name="heya.js"
    statuses={
      Object {
        "modified": false,
      }
    }
  />
</ul>
`;

exports[`FileSystem Component NewFileForm Component should display a duplicate warning if we try to enter the same filename twice 1`] = `
<div>
  <div
    className="input-group"
  >
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Filename"
      type="text"
      value=""
    />
    <span
      className="input-group-btn"
    >
      <button
        className="btn btn-secondary"
        onClick={[Function]}
        type="button"
      >
        Add File
      </button>
    </span>
  </div>
</div>
`;

exports[`FileSystem Component NewFileForm Component should render correctly 1`] = `
<div>
  <div
    className="input-group"
  >
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Filename"
      type="text"
      value=""
    />
    <span
      className="input-group-btn"
    >
      <button
        className="btn btn-secondary"
        onClick={[Function]}
        type="button"
      >
        Add File
      </button>
    </span>
  </div>
</div>
`;

exports[`FileSystem Component does not render form if file count is equal to max count 1`] = `
<div
  className="col-lg-3 col-sm-12"
>
  <h4>
    Files (Max of 
    5
    )
  </h4>
  <Component
    fileStatus={Object {}}
    files={
      Array [
        "a",
        "b",
        "c",
        "d",
        "e",
      ]
    }
  />
</div>
`;

exports[`FileSystem Component renders form if file count is less than max count 1`] = `
<div
  className="col-lg-3 col-sm-12"
>
  <h4>
    Files (Max of 
    5
    )
  </h4>
  <Component
    fileStatus={Object {}}
    files={Array []}
  />
  <NewFileForm
    files={Array []}
  />
</div>
`;
